// drizzle-orm@0.28.6/pg-core/index.cjs
'use strict';

var index = require('../index-1899b9ae.cjs');

function alias(table, alias) {
    return new Proxy(table, new index.TableAliasProxyHandler(alias, false));
}

exports.Check = index.Check;
exports.CheckBuilder = index.CheckBuilder;
exports.DefaultViewBuilderCore = index.DefaultViewBuilderCore;
exports.ForeignKey = index.ForeignKey;
exports.ForeignKeyBuilder = index.ForeignKeyBuilder;
exports.Index = index.Index;
exports.IndexBuilder = index.IndexBuilder;
exports.IndexBuilderOn = index.IndexBuilderOn;
exports.InlineForeignKeys = index.InlineForeignKeys;
exports.ManualMaterializedViewBuilder = index.ManualMaterializedViewBuilder;
exports.ManualViewBuilder = index.ManualViewBuilder;
exports.MaterializedViewBuilder = index.MaterializedViewBuilder;
exports.MaterializedViewBuilderCore = index.MaterializedViewBuilderCore;
exports.PgArray = index.PgArray;
exports.PgArrayBuilder = index.PgArrayBuilder;
exports.PgBigInt53 = index.PgBigInt53;
exports.PgBigInt53Builder = index.PgBigInt53Builder;
exports.PgBigInt64 = index.PgBigInt64;
exports.PgBigInt64Builder = index.PgBigInt64Builder;
exports.PgBigSerial53 = index.PgBigSerial53;
exports.PgBigSerial53Builder = index.PgBigSerial53Builder;
exports.PgBigSerial64 = index.PgBigSerial64;
exports.PgBigSerial64Builder = index.PgBigSerial64Builder;
exports.PgBoolean = index.PgBoolean;
exports.PgBooleanBuilder = index.PgBooleanBuilder;
exports.PgChar = index.PgChar;
exports.PgCharBuilder = index.PgCharBuilder;
exports.PgCidr = index.PgCidr;
exports.PgCidrBuilder = index.PgCidrBuilder;
exports.PgColumn = index.PgColumn;
exports.PgColumnBuilder = index.PgColumnBuilder;
exports.PgCustomColumn = index.PgCustomColumn;
exports.PgCustomColumnBuilder = index.PgCustomColumnBuilder;
exports.PgDatabase = index.PgDatabase;
exports.PgDate = index.PgDate;
exports.PgDateBuilder = index.PgDateBuilder;
exports.PgDateString = index.PgDateString;
exports.PgDateStringBuilder = index.PgDateStringBuilder;
exports.PgDelete = index.PgDelete;
exports.PgDialect = index.PgDialect;
exports.PgDoublePrecision = index.PgDoublePrecision;
exports.PgDoublePrecisionBuilder = index.PgDoublePrecisionBuilder;
exports.PgEnumColumn = index.PgEnumColumn;
exports.PgEnumColumnBuilder = index.PgEnumColumnBuilder;
exports.PgInet = index.PgInet;
exports.PgInetBuilder = index.PgInetBuilder;
exports.PgInsert = index.PgInsert;
exports.PgInsertBuilder = index.PgInsertBuilder;
exports.PgInteger = index.PgInteger;
exports.PgIntegerBuilder = index.PgIntegerBuilder;
exports.PgInterval = index.PgInterval;
exports.PgIntervalBuilder = index.PgIntervalBuilder;
exports.PgJson = index.PgJson;
exports.PgJsonBuilder = index.PgJsonBuilder;
exports.PgJsonb = index.PgJsonb;
exports.PgJsonbBuilder = index.PgJsonbBuilder;
exports.PgMacaddr = index.PgMacaddr;
exports.PgMacaddr8 = index.PgMacaddr8;
exports.PgMacaddr8Builder = index.PgMacaddr8Builder;
exports.PgMacaddrBuilder = index.PgMacaddrBuilder;
exports.PgMaterializedView = index.PgMaterializedView;
exports.PgMaterializedViewConfig = index.PgMaterializedViewConfig;
exports.PgNumeric = index.PgNumeric;
exports.PgNumericBuilder = index.PgNumericBuilder;
exports.PgReal = index.PgReal;
exports.PgRealBuilder = index.PgRealBuilder;
exports.PgRefreshMaterializedView = index.PgRefreshMaterializedView;
exports.PgSchema = index.PgSchema;
exports.PgSelect = index.PgSelect;
exports.PgSelectBuilder = index.PgSelectBuilder;
exports.PgSelectQueryBuilder = index.PgSelectQueryBuilder;
exports.PgSerial = index.PgSerial;
exports.PgSerialBuilder = index.PgSerialBuilder;
exports.PgSession = index.PgSession;
exports.PgSmallInt = index.PgSmallInt;
exports.PgSmallIntBuilder = index.PgSmallIntBuilder;
exports.PgSmallSerial = index.PgSmallSerial;
exports.PgSmallSerialBuilder = index.PgSmallSerialBuilder;
exports.PgTable = index.PgTable;
exports.PgText = index.PgText;
exports.PgTextBuilder = index.PgTextBuilder;
exports.PgTime = index.PgTime;
exports.PgTimeBuilder = index.PgTimeBuilder;
exports.PgTimestamp = index.PgTimestamp;
exports.PgTimestampBuilder = index.PgTimestampBuilder;
exports.PgTimestampString = index.PgTimestampString;
exports.PgTimestampStringBuilder = index.PgTimestampStringBuilder;
exports.PgTransaction = index.PgTransaction;
exports.PgUUID = index.PgUUID;
exports.PgUUIDBuilder = index.PgUUIDBuilder;
exports.PgUpdate = index.PgUpdate;
exports.PgUpdateBuilder = index.PgUpdateBuilder;
exports.PgVarchar = index.PgVarchar;
exports.PgVarcharBuilder = index.PgVarcharBuilder;
exports.PgView = index.PgView;
exports.PgViewBase = index.PgViewBase;
exports.PgViewConfig = index.PgViewConfig;
exports.PreparedQuery = index.PreparedQuery;
exports.PrimaryKey = index.PrimaryKey;
exports.PrimaryKeyBuilder = index.PrimaryKeyBuilder;
exports.QueryBuilder = index.QueryBuilder;
exports.UniqueConstraint = index.UniqueConstraint;
exports.UniqueConstraintBuilder = index.UniqueConstraintBuilder;
exports.UniqueOnConstraintBuilder = index.UniqueOnConstraintBuilder;
exports.ViewBuilder = index.ViewBuilder;
exports.bigint = index.bigint;
exports.bigserial = index.bigserial;
exports.boolean = index.boolean;
exports.char = index.char;
exports.check = index.check;
exports.cidr = index.cidr;
exports.customType = index.customType;
exports.date = index.date;
exports.decimal = index.decimal;
exports.doublePrecision = index.doublePrecision;
exports.foreignKey = index.foreignKey;
exports.getMaterializedViewConfig = index.getMaterializedViewConfig;
exports.getTableConfig = index.getTableConfig;
exports.getViewConfig = index.getViewConfig;
exports.index = index.index;
exports.inet = index.inet;
exports.integer = index.integer;
exports.interval = index.interval;
exports.isPgEnum = index.isPgEnum;
exports.isPgSchema = index.isPgSchema;
exports.json = index.json;
exports.jsonb = index.jsonb;
exports.macaddr = index.macaddr;
exports.macaddr8 = index.macaddr8;
exports.makePgArray = index.makePgArray;
exports.numeric = index.numeric;
exports.parsePgArray = index.parsePgArray;
exports.parsePgNestedArray = index.parsePgNestedArray;
exports.pgEnum = index.pgEnum;
exports.pgMaterializedView = index.pgMaterializedView;
exports.pgMaterializedViewWithSchema = index.pgMaterializedViewWithSchema;
exports.pgSchema = index.pgSchema;
exports.pgTable = index.pgTable;
exports.pgTableCreator = index.pgTableCreator;
exports.pgTableWithSchema = index.pgTableWithSchema;
exports.pgView = index.pgView;
exports.pgViewWithSchema = index.pgViewWithSchema;
exports.primaryKey = index.primaryKey;
exports.real = index.real;
exports.serial = index.serial;
exports.smallint = index.smallint;
exports.smallserial = index.smallserial;
exports.text = index.text;
exports.time = index.time;
exports.timestamp = index.timestamp;
exports.unique = index.unique;
exports.uniqueIndex = index.uniqueIndex;
exports.uniqueKeyName = index.uniqueKeyName;
exports.uuid = index.uuid;
exports.varchar = index.varchar;
exports.alias = alias;
