// drizzle-orm@0.28.6/mysql-core/index.mjs
import { e as entityKind, s as sql, n as TableAliasProxyHandler, i as is, T as Table, V as ViewBaseConfig } from '../alias-cf8e03cd.mjs';
import { d as MySqlColumnBuilderWithAutoIncrement, e as MySqlColumnWithAutoIncrement, f as MySqlColumnBuilder, g as MySqlColumn, h as MySqlTable, m as mysqlTableWithSchema, i as mysqlViewWithSchema, U as UniqueConstraintBuilder, F as ForeignKeyBuilder, j as MySqlViewConfig } from '../view-23898f21.mjs';
export { k as ForeignKey, I as InlineForeignKeys, C as ManualViewBuilder, b as MySqlDatabase, n as MySqlDelete, c as MySqlDialect, p as MySqlInsert, o as MySqlInsertBuilder, s as MySqlSelect, q as MySqlSelectBuilder, r as MySqlSelectQueryBuilder, M as MySqlSession, a as MySqlTransaction, u as MySqlUpdate, t as MySqlUpdateBuilder, E as MySqlView, D as MySqlViewBase, P as PreparedQuery, Q as QueryBuilder, A as UniqueConstraint, z as UniqueOnConstraintBuilder, B as ViewBuilder, V as ViewBuilderCore, l as foreignKey, v as mysqlTable, w as mysqlTableCreator, G as mysqlView, x as unique, y as uniqueKeyName } from '../view-23898f21.mjs';

class MySqlBigInt53Builder extends MySqlColumnBuilderWithAutoIncrement {
    static [entityKind] = 'MySqlBigInt53Builder';
    constructor(name) {
        super(name, 'number', 'MySqlBigInt53');
    }
    /** @internal */
    build(table) {
        return new MySqlBigInt53(table, this.config);
    }
}
class MySqlBigInt53 extends MySqlColumnWithAutoIncrement {
    static [entityKind] = 'MySqlBigInt53';
    getSQLType() {
        return 'bigint';
    }
    mapFromDriverValue(value) {
        if (typeof value === 'number') {
            return value;
        }
        return Number(value);
    }
}
class MySqlBigInt64Builder extends MySqlColumnBuilderWithAutoIncrement {
    static [entityKind] = 'MySqlBigInt64Builder';
    constructor(name) {
        super(name, 'bigint', 'MySqlBigInt64');
    }
    /** @internal */
    build(table) {
        return new MySqlBigInt64(table, this.config);
    }
}
class MySqlBigInt64 extends MySqlColumnWithAutoIncrement {
    static [entityKind] = 'MySqlBigInt64';
    getSQLType() {
        return 'bigint';
    }
    // eslint-disable-next-line unicorn/prefer-native-coercion-functions
    mapFromDriverValue(value) {
        return BigInt(value);
    }
}
function bigint(name, config) {
    if (config.mode === 'number') {
        return new MySqlBigInt53Builder(name);
    }
    return new MySqlBigInt64Builder(name);
}

class MySqlBinaryBuilder extends MySqlColumnBuilder {
    static [entityKind] = 'MySqlBinaryBuilder';
    constructor(name, length) {
        super(name, 'string', 'MySqlBinary');
        this.config.length = length;
    }
    /** @internal */
    build(table) {
        return new MySqlBinary(table, this.config);
    }
}
class MySqlBinary extends MySqlColumn {
    static [entityKind] = 'MySqlBinary';
    length = this.config.length;
    getSQLType() {
        return this.length === undefined ? `binary` : `binary(${this.length})`;
    }
}
function binary(name, config = {}) {
    return new MySqlBinaryBuilder(name, config.length);
}

class MySqlBooleanBuilder extends MySqlColumnBuilder {
    static [entityKind] = 'MySqlBooleanBuilder';
    constructor(name) {
        super(name, 'boolean', 'MySqlBoolean');
    }
    /** @internal */
    build(table) {
        return new MySqlBoolean(table, this.config);
    }
}
class MySqlBoolean extends MySqlColumn {
    static [entityKind] = 'MySqlBoolean';
    getSQLType() {
        return 'boolean';
    }
    mapFromDriverValue(value) {
        if (typeof value === 'boolean') {
            return value;
        }
        return value === 1;
    }
}
function boolean(name) {
    return new MySqlBooleanBuilder(name);
}

class MySqlCharBuilder extends MySqlColumnBuilder {
    static [entityKind] = 'MySqlCharBuilder';
    constructor(name, config) {
        super(name, 'string', 'MySqlChar');
        this.config.length = config.length;
        this.config.enum = config.enum;
    }
    /** @internal */
    build(table) {
        return new MySqlChar(table, this.config);
    }
}
class MySqlChar extends MySqlColumn {
    static [entityKind] = 'MySqlChar';
    length = this.config.length;
    enumValues = this.config.enum;
    getSQLType() {
        return this.length === undefined ? `char` : `char(${this.length})`;
    }
}
function char(name, config = {}) {
    return new MySqlCharBuilder(name, config);
}

class MySqlCustomColumnBuilder extends MySqlColumnBuilder {
    static [entityKind] = 'MySqlCustomColumnBuilder';
    constructor(name, fieldConfig, customTypeParams) {
        super(name, 'custom', 'MySqlCustomColumn');
        this.config.fieldConfig = fieldConfig;
        this.config.customTypeParams = customTypeParams;
    }
    /** @internal */
    build(table) {
        return new MySqlCustomColumn(table, this.config);
    }
}
class MySqlCustomColumn extends MySqlColumn {
    static [entityKind] = 'MySqlCustomColumn';
    sqlName;
    mapTo;
    mapFrom;
    constructor(table, config) {
        super(table, config);
        this.sqlName = config.customTypeParams.dataType(config.fieldConfig);
        this.mapTo = config.customTypeParams.toDriver;
        this.mapFrom = config.customTypeParams.fromDriver;
    }
    getSQLType() {
        return this.sqlName;
    }
    mapFromDriverValue(value) {
        return typeof this.mapFrom === 'function' ? this.mapFrom(value) : value;
    }
    mapToDriverValue(value) {
        return typeof this.mapTo === 'function' ? this.mapTo(value) : value;
    }
}
/**
 * Custom mysql database data type generator
 */
function customType(customTypeParams) {
    return (dbName, fieldConfig) => {
        return new MySqlCustomColumnBuilder(dbName, fieldConfig, customTypeParams);
    };
}

class MySqlDateBuilder extends MySqlColumnBuilder {
    static [entityKind] = 'MySqlDateBuilder';
    constructor(name) {
        super(name, 'date', 'MySqlDate');
    }
    /** @internal */
    build(table) {
        return new MySqlDate(table, this.config);
    }
}
class MySqlDate extends MySqlColumn {
    static [entityKind] = 'MySqlDate';
    constructor(table, config) {
        super(table, config);
    }
    getSQLType() {
        return `date`;
    }
    mapFromDriverValue(value) {
        return new Date(value);
    }
}
class MySqlDateStringBuilder extends MySqlColumnBuilder {
    static [entityKind] = 'MySqlDateStringBuilder';
    constructor(name) {
        super(name, 'string', 'MySqlDateString');
    }
    /** @internal */
    build(table) {
        return new MySqlDateString(table, this.config);
    }
}
class MySqlDateString extends MySqlColumn {
    static [entityKind] = 'MySqlDateString';
    constructor(table, config) {
        super(table, config);
    }
    getSQLType() {
        return `date`;
    }
}
function date(name, config = {}) {
    if (config.mode === 'string') {
        return new MySqlDateStringBuilder(name);
    }
    return new MySqlDateBuilder(name);
}

class MySqlDateTimeBuilder extends MySqlColumnBuilder {
    static [entityKind] = 'MySqlDateTimeBuilder';
    constructor(name, config) {
        super(name, 'date', 'MySqlDateTime');
        this.config.fsp = config?.fsp;
    }
    /** @internal */
    build(table) {
        return new MySqlDateTime(table, this.config);
    }
}
class MySqlDateTime extends MySqlColumn {
    static [entityKind] = 'MySqlDateTime';
    fsp;
    constructor(table, config) {
        super(table, config);
        this.fsp = config.fsp;
    }
    getSQLType() {
        const precision = this.fsp === undefined ? '' : `(${this.fsp})`;
        return `datetime${precision}`;
    }
    mapToDriverValue(value) {
        return value.toISOString().replace('T', ' ').replace('Z', '');
    }
    mapFromDriverValue(value) {
        return new Date(value.replace(' ', 'T') + 'Z');
    }
}
class MySqlDateTimeStringBuilder extends MySqlColumnBuilder {
    static [entityKind] = 'MySqlDateTimeStringBuilder';
    constructor(name, config) {
        super(name, 'string', 'MySqlDateTimeString');
        this.config.fsp = config?.fsp;
    }
    /** @internal */
    build(table) {
        return new MySqlDateTimeString(table, this.config);
    }
}
class MySqlDateTimeString extends MySqlColumn {
    static [entityKind] = 'MySqlDateTimeString';
    fsp;
    constructor(table, config) {
        super(table, config);
        this.fsp = config.fsp;
    }
    getSQLType() {
        const precision = this.fsp === undefined ? '' : `(${this.fsp})`;
        return `datetime${precision}`;
    }
}
function datetime(name, config = {}) {
    if (config.mode === 'string') {
        return new MySqlDateTimeStringBuilder(name, config);
    }
    return new MySqlDateTimeBuilder(name, config);
}

class MySqlDecimalBuilder extends MySqlColumnBuilderWithAutoIncrement {
    static [entityKind] = 'MySqlDecimalBuilder';
    constructor(name, precision, scale) {
        super(name, 'string', 'MySqlDecimal');
        this.config.precision = precision;
        this.config.scale = scale;
    }
    /** @internal */
    build(table) {
        return new MySqlDecimal(table, this.config);
    }
}
class MySqlDecimal extends MySqlColumnWithAutoIncrement {
    static [entityKind] = 'MySqlDecimal';
    precision = this.config.precision;
    scale = this.config.scale;
    getSQLType() {
        if (this.precision !== undefined && this.scale !== undefined) {
            return `decimal(${this.precision},${this.scale})`;
        }
        else if (this.precision === undefined) {
            return 'decimal';
        }
        else {
            return `decimal(${this.precision})`;
        }
    }
}
function decimal(name, config = {}) {
    return new MySqlDecimalBuilder(name, config.precision, config.scale);
}

class MySqlDoubleBuilder extends MySqlColumnBuilderWithAutoIncrement {
    static [entityKind] = 'MySqlDoubleBuilder';
    constructor(name, config) {
        super(name, 'number', 'MySqlDouble');
        this.config.precision = config?.precision;
        this.config.scale = config?.scale;
    }
    /** @internal */
    build(table) {
        return new MySqlDouble(table, this.config);
    }
}
class MySqlDouble extends MySqlColumnWithAutoIncrement {
    static [entityKind] = 'MySqlDouble';
    precision = this.config.precision;
    scale = this.config.scale;
    getSQLType() {
        if (this.precision !== undefined && this.scale !== undefined) {
            return `double(${this.precision},${this.scale})`;
        }
        else if (this.precision === undefined) {
            return 'double';
        }
        else {
            return `double(${this.precision})`;
        }
    }
}
function double(name, config) {
    return new MySqlDoubleBuilder(name, config);
}

class MySqlEnumColumnBuilder extends MySqlColumnBuilder {
    static [entityKind] = 'MySqlEnumColumnBuilder';
    constructor(name, values) {
        super(name, 'string', 'MySqlEnumColumn');
        this.config.enumValues = values;
    }
    /** @internal */
    build(table) {
        return new MySqlEnumColumn(table, this.config);
    }
}
class MySqlEnumColumn extends MySqlColumn {
    static [entityKind] = 'MySqlEnumColumn';
    enumValues = this.config.enumValues;
    getSQLType() {
        return `enum(${this.enumValues.map((value) => `'${value}'`).join(',')})`;
    }
}
function mysqlEnum(name, values) {
    if (values.length === 0) {
        throw new Error(`You have an empty array for "${name}" enum values`);
    }
    return new MySqlEnumColumnBuilder(name, values);
}

class MySqlFloatBuilder extends MySqlColumnBuilderWithAutoIncrement {
    static [entityKind] = 'MySqlFloatBuilder';
    constructor(name) {
        super(name, 'number', 'MySqlFloat');
    }
    /** @internal */
    build(table) {
        return new MySqlFloat(table, this.config);
    }
}
class MySqlFloat extends MySqlColumnWithAutoIncrement {
    static [entityKind] = 'MySqlFloat';
    getSQLType() {
        return 'float';
    }
}
function float(name) {
    return new MySqlFloatBuilder(name);
}

class MySqlIntBuilder extends MySqlColumnBuilderWithAutoIncrement {
    static [entityKind] = 'MySqlIntBuilder';
    constructor(name) {
        super(name, 'number', 'MySqlInt');
    }
    /** @internal */
    build(table) {
        return new MySqlInt(table, this.config);
    }
}
class MySqlInt extends MySqlColumnWithAutoIncrement {
    static [entityKind] = 'MySqlInt';
    getSQLType() {
        return 'int';
    }
    mapFromDriverValue(value) {
        if (typeof value === 'string') {
            return Number(value);
        }
        return value;
    }
}
function int(name) {
    return new MySqlIntBuilder(name);
}

class MySqlJsonBuilder extends MySqlColumnBuilder {
    static [entityKind] = 'MySqlJsonBuilder';
    constructor(name) {
        super(name, 'json', 'MySqlJson');
    }
    /** @internal */
    build(table) {
        return new MySqlJson(table, this.config);
    }
}
class MySqlJson extends MySqlColumn {
    static [entityKind] = 'MySqlJson';
    getSQLType() {
        return 'json';
    }
    mapToDriverValue(value) {
        return JSON.stringify(value);
    }
}
function json(name) {
    return new MySqlJsonBuilder(name);
}

class MySqlMediumIntBuilder extends MySqlColumnBuilderWithAutoIncrement {
    static [entityKind] = 'MySqlMediumIntBuilder';
    constructor(name) {
        super(name, 'number', 'MySqlMediumInt');
    }
    /** @internal */
    build(table) {
        return new MySqlMediumInt(table, this.config);
    }
}
class MySqlMediumInt extends MySqlColumnWithAutoIncrement {
    static [entityKind] = 'MySqlMediumInt';
    getSQLType() {
        return 'mediumint';
    }
    mapFromDriverValue(value) {
        if (typeof value === 'string') {
            return Number(value);
        }
        return value;
    }
}
function mediumint(name) {
    return new MySqlMediumIntBuilder(name);
}

class MySqlRealBuilder extends MySqlColumnBuilderWithAutoIncrement {
    static [entityKind] = 'MySqlRealBuilder';
    constructor(name, config) {
        super(name, 'number', 'MySqlReal');
        this.config.precision = config?.precision;
        this.config.scale = config?.scale;
    }
    /** @internal */
    build(table) {
        return new MySqlReal(table, this.config);
    }
}
class MySqlReal extends MySqlColumnWithAutoIncrement {
    static [entityKind] = 'MySqlReal';
    precision = this.config.precision;
    scale = this.config.scale;
    getSQLType() {
        if (this.precision !== undefined && this.scale !== undefined) {
            return `real(${this.precision}, ${this.scale})`;
        }
        else if (this.precision === undefined) {
            return 'real';
        }
        else {
            return `real(${this.precision})`;
        }
    }
}
function real(name, config = {}) {
    return new MySqlRealBuilder(name, config);
}

class MySqlSerialBuilder extends MySqlColumnBuilderWithAutoIncrement {
    static [entityKind] = 'MySqlSerialBuilder';
    constructor(name) {
        super(name, 'number', 'MySqlSerial');
        this.config.hasDefault = true;
        this.config.autoIncrement = true;
    }
    /** @internal */
    build(table) {
        return new MySqlSerial(table, this.config);
    }
}
class MySqlSerial extends MySqlColumnWithAutoIncrement {
    static [entityKind] = 'MySqlSerial';
    getSQLType() {
        return 'serial';
    }
    mapFromDriverValue(value) {
        if (typeof value === 'string') {
            return Number(value);
        }
        return value;
    }
}
function serial(name) {
    return new MySqlSerialBuilder(name);
}

class MySqlSmallIntBuilder extends MySqlColumnBuilderWithAutoIncrement {
    static [entityKind] = 'MySqlSmallIntBuilder';
    constructor(name) {
        super(name, 'number', 'MySqlSmallInt');
    }
    /** @internal */
    build(table) {
        return new MySqlSmallInt(table, this.config);
    }
}
class MySqlSmallInt extends MySqlColumnWithAutoIncrement {
    static [entityKind] = 'MySqlSmallInt';
    getSQLType() {
        return 'smallint';
    }
    mapFromDriverValue(value) {
        if (typeof value === 'string') {
            return Number(value);
        }
        return value;
    }
}
function smallint(name) {
    return new MySqlSmallIntBuilder(name);
}

class MySqlTextBuilder extends MySqlColumnBuilder {
    static [entityKind] = 'MySqlTextBuilder';
    constructor(name, textType, config) {
        super(name, 'string', 'MySqlText');
        this.config.textType = textType;
        this.config.enumValues = config.enum;
    }
    /** @internal */
    build(table) {
        return new MySqlText(table, this.config);
    }
}
class MySqlText extends MySqlColumn {
    static [entityKind] = 'MySqlText';
    textType = this.config.textType;
    enumValues = this.config.enumValues;
    getSQLType() {
        return this.textType;
    }
}
function text(name, config = {}) {
    return new MySqlTextBuilder(name, 'text', config);
}
function tinytext(name, config = {}) {
    return new MySqlTextBuilder(name, 'tinytext', config);
}
function mediumtext(name, config = {}) {
    return new MySqlTextBuilder(name, 'mediumtext', config);
}
function longtext(name, config = {}) {
    return new MySqlTextBuilder(name, 'longtext', config);
}

class MySqlTimeBuilder extends MySqlColumnBuilder {
    static [entityKind] = 'MySqlTimeBuilder';
    constructor(name, config) {
        super(name, 'string', 'MySqlTime');
        this.config.fsp = config?.fsp;
    }
    /** @internal */
    build(table) {
        return new MySqlTime(table, this.config);
    }
}
class MySqlTime extends MySqlColumn {
    static [entityKind] = 'MySqlTime';
    fsp = this.config.fsp;
    getSQLType() {
        const precision = this.fsp === undefined ? '' : `(${this.fsp})`;
        return `time${precision}`;
    }
}
function time(name, config) {
    return new MySqlTimeBuilder(name, config);
}

class MySqlDateColumnBaseBuilder extends MySqlColumnBuilder {
    static [entityKind] = 'MySqlDateColumnBuilder';
    defaultNow() {
        return this.default(sql `(now())`);
    }
    // "on update now" also adds an implicit default value to the column - https://dev.mysql.com/doc/refman/8.0/en/timestamp-initialization.html
    onUpdateNow() {
        this.config.hasOnUpdateNow = true;
        this.config.hasDefault = true;
        return this;
    }
}
class MySqlDateBaseColumn extends MySqlColumn {
    static [entityKind] = 'MySqlDateColumn';
    hasOnUpdateNow = this.config.hasOnUpdateNow;
}

class MySqlTimestampBuilder extends MySqlDateColumnBaseBuilder {
    static [entityKind] = 'MySqlTimestampBuilder';
    constructor(name, config) {
        super(name, 'date', 'MySqlTimestamp');
        this.config.fsp = config?.fsp;
    }
    /** @internal */
    build(table) {
        return new MySqlTimestamp(table, this.config);
    }
}
class MySqlTimestamp extends MySqlDateBaseColumn {
    static [entityKind] = 'MySqlTimestamp';
    fsp = this.config.fsp;
    getSQLType() {
        const precision = this.fsp === undefined ? '' : `(${this.fsp})`;
        return `timestamp${precision}`;
    }
    mapFromDriverValue(value) {
        return new Date(value + '+0000');
    }
    mapToDriverValue(value) {
        return value.toISOString().slice(0, -1).replace('T', ' ');
    }
}
class MySqlTimestampStringBuilder extends MySqlDateColumnBaseBuilder {
    static [entityKind] = 'MySqlTimestampStringBuilder';
    constructor(name, config) {
        super(name, 'string', 'MySqlTimestampString');
        this.config.fsp = config?.fsp;
    }
    /** @internal */
    build(table) {
        return new MySqlTimestampString(table, this.config);
    }
}
class MySqlTimestampString extends MySqlDateBaseColumn {
    static [entityKind] = 'MySqlTimestampString';
    fsp = this.config.fsp;
    getSQLType() {
        const precision = this.fsp === undefined ? '' : `(${this.fsp})`;
        return `timestamp${precision}`;
    }
}
function timestamp(name, config = {}) {
    if (config.mode === 'string') {
        return new MySqlTimestampStringBuilder(name, config);
    }
    return new MySqlTimestampBuilder(name, config);
}

class MySqlTinyIntBuilder extends MySqlColumnBuilderWithAutoIncrement {
    static [entityKind] = 'MySqlTinyIntBuilder';
    constructor(name) {
        super(name, 'number', 'MySqlTinyInt');
    }
    /** @internal */
    build(table) {
        return new MySqlTinyInt(table, this.config);
    }
}
class MySqlTinyInt extends MySqlColumnWithAutoIncrement {
    static [entityKind] = 'MySqlTinyInt';
    getSQLType() {
        return 'tinyint';
    }
    mapFromDriverValue(value) {
        if (typeof value === 'string') {
            return Number(value);
        }
        return value;
    }
}
function tinyint(name) {
    return new MySqlTinyIntBuilder(name);
}

class MySqlVarBinaryBuilder extends MySqlColumnBuilder {
    static [entityKind] = 'MySqlVarBinaryBuilder';
    /** @internal */
    constructor(name, config) {
        super(name, 'string', 'MySqlVarBinary');
        this.config.length = config?.length;
    }
    /** @internal */
    build(table) {
        return new MySqlVarBinary(table, this.config);
    }
}
class MySqlVarBinary extends MySqlColumn {
    static [entityKind] = 'MySqlVarBinary';
    length = this.config.length;
    getSQLType() {
        return this.length === undefined ? `varbinary` : `varbinary(${this.length})`;
    }
}
function varbinary(name, options) {
    return new MySqlVarBinaryBuilder(name, options);
}

class MySqlVarCharBuilder extends MySqlColumnBuilder {
    static [entityKind] = 'MySqlVarCharBuilder';
    /** @internal */
    constructor(name, config) {
        super(name, 'string', 'MySqlVarChar');
        this.config.length = config.length;
        this.config.enum = config.enum;
    }
    /** @internal */
    build(table) {
        return new MySqlVarChar(table, this.config);
    }
}
class MySqlVarChar extends MySqlColumn {
    static [entityKind] = 'MySqlVarChar';
    length = this.config.length;
    enumValues = this.config.enum;
    getSQLType() {
        return this.length === undefined ? `varchar` : `varchar(${this.length})`;
    }
}
function varchar(name, config) {
    return new MySqlVarCharBuilder(name, config);
}

class MySqlYearBuilder extends MySqlColumnBuilder {
    static [entityKind] = 'MySqlYearBuilder';
    constructor(name) {
        super(name, 'number', 'MySqlYear');
    }
    /** @internal */
    build(table) {
        return new MySqlYear(table, this.config);
    }
}
class MySqlYear extends MySqlColumn {
    static [entityKind] = 'MySqlYear';
    getSQLType() {
        return `year`;
    }
}
function year(name) {
    return new MySqlYearBuilder(name);
}

function alias(table, alias) {
    return new Proxy(table, new TableAliasProxyHandler(alias, false));
}

class CheckBuilder {
    name;
    value;
    static [entityKind] = 'MySqlCheckBuilder';
    brand;
    constructor(name, value) {
        this.name = name;
        this.value = value;
    }
    /** @internal */
    build(table) {
        return new Check(table, this);
    }
}
class Check {
    table;
    static [entityKind] = 'MySqlCheck';
    name;
    value;
    constructor(table, builder) {
        this.table = table;
        this.name = builder.name;
        this.value = builder.value;
    }
}
function check(name, value) {
    return new CheckBuilder(name, value);
}

class IndexBuilderOn {
    name;
    unique;
    static [entityKind] = 'MySqlIndexBuilderOn';
    constructor(name, unique) {
        this.name = name;
        this.unique = unique;
    }
    on(...columns) {
        return new IndexBuilder(this.name, columns, this.unique);
    }
}
class IndexBuilder {
    static [entityKind] = 'MySqlIndexBuilder';
    /** @internal */
    config;
    constructor(name, columns, unique) {
        this.config = {
            name,
            columns,
            unique,
        };
    }
    using(using) {
        this.config.using = using;
        return this;
    }
    algorythm(algorythm) {
        this.config.algorythm = algorythm;
        return this;
    }
    lock(lock) {
        this.config.lock = lock;
        return this;
    }
    /** @internal */
    build(table) {
        return new Index(this.config, table);
    }
}
class Index {
    static [entityKind] = 'MySqlIndex';
    config;
    constructor(config, table) {
        this.config = { ...config, table };
    }
}
function index(name) {
    return new IndexBuilderOn(name, false);
}
function uniqueIndex(name) {
    return new IndexBuilderOn(name, true);
}

function primaryKey(...columns) {
    return new PrimaryKeyBuilder(columns);
}
class PrimaryKeyBuilder {
    static [entityKind] = 'MySqlPrimaryKeyBuilder';
    /** @internal */
    columns;
    constructor(columns) {
        this.columns = columns;
    }
    /** @internal */
    build(table) {
        return new PrimaryKey(table, this.columns);
    }
}
class PrimaryKey {
    table;
    static [entityKind] = 'MySqlPrimaryKey';
    columns;
    constructor(table, columns) {
        this.table = table;
        this.columns = columns;
    }
    getName() {
        return `${this.table[MySqlTable.Symbol.Name]}_${this.columns.map((column) => column.name).join('_')}_pk`;
    }
}

class MySqlSchema {
    schemaName;
    static [entityKind] = 'MySqlSchema';
    constructor(schemaName) {
        this.schemaName = schemaName;
    }
    table = (name, columns, extraConfig) => {
        return mysqlTableWithSchema(name, columns, extraConfig, this.schemaName);
    };
    view = ((name, columns) => {
        return mysqlViewWithSchema(name, columns, this.schemaName);
    });
}
/** @deprecated - use `instanceof MySqlSchema` */
function isMySqlSchema(obj) {
    return is(obj, MySqlSchema);
}
/**
 * Create a MySQL schema.
 * https://dev.mysql.com/doc/refman/8.0/en/create-database.html
 *
 * @param name mysql use schema name
 * @returns MySQL schema
 */
function mysqlDatabase(name) {
    return new MySqlSchema(name);
}
/**
 * @see mysqlDatabase
 */
const mysqlSchema = mysqlDatabase;

function getTableConfig(table) {
    const columns = Object.values(table[MySqlTable.Symbol.Columns]);
    const indexes = [];
    const checks = [];
    const primaryKeys = [];
    const uniqueConstraints = [];
    const foreignKeys = Object.values(table[MySqlTable.Symbol.InlineForeignKeys]);
    const name = table[Table.Symbol.Name];
    const schema = table[Table.Symbol.Schema];
    const baseName = table[Table.Symbol.BaseName];
    const extraConfigBuilder = table[MySqlTable.Symbol.ExtraConfigBuilder];
    if (extraConfigBuilder !== undefined) {
        const extraConfig = extraConfigBuilder(table[MySqlTable.Symbol.Columns]);
        for (const builder of Object.values(extraConfig)) {
            if (is(builder, IndexBuilder)) {
                indexes.push(builder.build(table));
            }
            else if (is(builder, CheckBuilder)) {
                checks.push(builder.build(table));
            }
            else if (is(builder, UniqueConstraintBuilder)) {
                uniqueConstraints.push(builder.build(table));
            }
            else if (is(builder, PrimaryKeyBuilder)) {
                primaryKeys.push(builder.build(table));
            }
            else if (is(builder, ForeignKeyBuilder)) {
                foreignKeys.push(builder.build(table));
            }
        }
    }
    return {
        columns,
        indexes,
        foreignKeys,
        checks,
        primaryKeys,
        uniqueConstraints,
        name,
        schema,
        baseName,
    };
}
function getViewConfig(view) {
    return {
        ...view[ViewBaseConfig],
        ...view[MySqlViewConfig],
    };
}

export { Check, CheckBuilder, ForeignKeyBuilder, Index, IndexBuilder, IndexBuilderOn, MySqlBigInt53, MySqlBigInt53Builder, MySqlBigInt64, MySqlBigInt64Builder, MySqlBinary, MySqlBinaryBuilder, MySqlBoolean, MySqlBooleanBuilder, MySqlChar, MySqlCharBuilder, MySqlColumn, MySqlColumnBuilder, MySqlColumnBuilderWithAutoIncrement, MySqlColumnWithAutoIncrement, MySqlCustomColumn, MySqlCustomColumnBuilder, MySqlDate, MySqlDateBuilder, MySqlDateString, MySqlDateStringBuilder, MySqlDateTime, MySqlDateTimeBuilder, MySqlDateTimeString, MySqlDateTimeStringBuilder, MySqlDecimal, MySqlDecimalBuilder, MySqlDouble, MySqlDoubleBuilder, MySqlEnumColumn, MySqlEnumColumnBuilder, MySqlFloat, MySqlFloatBuilder, MySqlInt, MySqlIntBuilder, MySqlJson, MySqlJsonBuilder, MySqlMediumInt, MySqlMediumIntBuilder, MySqlReal, MySqlRealBuilder, MySqlSchema, MySqlSerial, MySqlSerialBuilder, MySqlSmallInt, MySqlSmallIntBuilder, MySqlTable, MySqlText, MySqlTextBuilder, MySqlTime, MySqlTimeBuilder, MySqlTimestamp, MySqlTimestampBuilder, MySqlTimestampString, MySqlTimestampStringBuilder, MySqlTinyInt, MySqlTinyIntBuilder, MySqlVarBinary, MySqlVarBinaryBuilder, MySqlVarChar, MySqlVarCharBuilder, MySqlViewConfig, MySqlYear, MySqlYearBuilder, PrimaryKey, PrimaryKeyBuilder, UniqueConstraintBuilder, alias, bigint, binary, boolean, char, check, customType, date, datetime, decimal, double, float, getTableConfig, getViewConfig, index, int, isMySqlSchema, json, longtext, mediumint, mediumtext, mysqlDatabase, mysqlEnum, mysqlSchema, mysqlTableWithSchema, mysqlViewWithSchema, primaryKey, real, serial, smallint, text, time, timestamp, tinyint, tinytext, uniqueIndex, varbinary, varchar, year };
