// drizzle-orm@0.28.6/neon-http/index.cjs
'use strict';

var serverless = require('@neondatabase/serverless');
var index = require('../index-1899b9ae.cjs');

var _a$1, _b, _c;
class NeonHttpPreparedQuery extends index.PreparedQuery {
    constructor(client, queryString, params, logger, fields, name, customResultMapper) {
        super();
        this.client = client;
        this.queryString = queryString;
        this.params = params;
        this.logger = logger;
        this.fields = fields;
        this.name = name;
        this.customResultMapper = customResultMapper;
        this.rawQuery = {
            arrayMode: false,
            fullResults: true,
        };
        this.query = { arrayMode: true, fullResults: true };
    }
    async execute(placeholderValues = {}) {
        const params = index.fillPlaceholders(this.params, placeholderValues);
        this.logger.logQuery(this.queryString, params);
        const { fields, client, queryString, query, rawQuery, joinsNotNullableMap, customResultMapper } = this;
        if (!fields && !customResultMapper) {
            return client(queryString, params, rawQuery);
        }
        const result = await client(queryString, params, query);
        return customResultMapper
            ? customResultMapper(result.rows)
            : result.rows.map((row) => index.mapResultRow(fields, row, joinsNotNullableMap));
    }
    all(placeholderValues = {}) {
        const params = index.fillPlaceholders(this.params, placeholderValues);
        this.logger.logQuery(this.queryString, params);
        return this.client(this.queryString, params, this.rawQuery).then((result) => result.rows);
    }
    values(placeholderValues = {}) {
        const params = index.fillPlaceholders(this.params, placeholderValues);
        this.logger.logQuery(this.queryString, params);
        return this.client(this.queryString, params).then((result) => result.rows);
    }
}
_a$1 = index.entityKind;
NeonHttpPreparedQuery[_a$1] = 'NeonHttpPreparedQuery';
class NeonHttpSession extends index.PgSession {
    constructor(client, dialect, schema, options = {}) {
        super(dialect);
        this.client = client;
        this.schema = schema;
        this.options = options;
        this.logger = options.logger ?? new index.NoopLogger();
    }
    prepareQuery(query, fields, name, customResultMapper) {
        return new NeonHttpPreparedQuery(this.client, query.sql, query.params, this.logger, fields, name, customResultMapper);
    }
    // change return type to QueryRows<true>
    async query(query, params) {
        this.logger.logQuery(query, params);
        const result = await this.client(query, params, { arrayMode: true });
        return result;
    }
    // change return type to QueryRows<false>
    async queryObjects(query, params) {
        return this.client(query, params);
    }
    async transaction(_transaction, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    _config = {}) {
        throw new Error('No transactions support in neon-http driver');
    }
}
_b = index.entityKind;
NeonHttpSession[_b] = 'NeonHttpSession';
class NeonTransaction extends index.PgTransaction {
    async transaction(_transaction) {
        throw new Error('No transactions support in neon-http driver');
        // const savepointName = `sp${this.nestedIndex + 1}`;
        // const tx = new NeonTransaction(this.dialect, this.session, this.schema, this.nestedIndex + 1);
        // await tx.execute(sql.raw(`savepoint ${savepointName}`));
        // try {
        // 	const result = await transaction(tx);
        // 	await tx.execute(sql.raw(`release savepoint ${savepointName}`));
        // 	return result;
        // } catch (e) {
        // 	await tx.execute(sql.raw(`rollback to savepoint ${savepointName}`));
        // 	throw e;
        // }
    }
}
_c = index.entityKind;
NeonTransaction[_c] = 'NeonHttpTransaction';

var _a;
class NeonHttpDriver {
    constructor(client, dialect, options = {}) {
        this.client = client;
        this.dialect = dialect;
        this.options = options;
        this.initMappers();
    }
    createSession(schema) {
        return new NeonHttpSession(this.client, this.dialect, schema, { logger: this.options.logger });
    }
    initMappers() {
        serverless.types.setTypeParser(serverless.types.builtins.TIMESTAMPTZ, (val) => val);
        serverless.types.setTypeParser(serverless.types.builtins.TIMESTAMP, (val) => val);
        serverless.types.setTypeParser(serverless.types.builtins.DATE, (val) => val);
    }
}
_a = index.entityKind;
NeonHttpDriver[_a] = 'NeonDriver';
function drizzle(client, config = {}) {
    const dialect = new index.PgDialect();
    let logger;
    if (config.logger === true) {
        logger = new index.DefaultLogger();
    }
    else if (config.logger !== false) {
        logger = config.logger;
    }
    let schema;
    if (config.schema) {
        const tablesConfig = index.extractTablesRelationalConfig(config.schema, index.createTableRelationsHelpers);
        schema = {
            fullSchema: config.schema,
            schema: tablesConfig.tables,
            tableNamesMap: tablesConfig.tableNamesMap,
        };
    }
    const driver = new NeonHttpDriver(client, dialect, { logger });
    const session = driver.createSession(schema);
    return new index.PgDatabase(dialect, session, schema);
}

exports.NeonHttpDriver = NeonHttpDriver;
exports.NeonHttpPreparedQuery = NeonHttpPreparedQuery;
exports.NeonHttpSession = NeonHttpSession;
exports.NeonTransaction = NeonTransaction;
exports.drizzle = drizzle;
