// drizzle-orm@0.28.6/sql-js/index.cjs
'use strict';

var index = require('../index-1899b9ae.cjs');
var session = require('../session-90f70450.cjs');

var _a, _b, _c;
class SQLJsSession extends session.SQLiteSession {
    constructor(client, dialect, schema, options = {}) {
        super(dialect);
        this.client = client;
        this.schema = schema;
        this.logger = options.logger ?? new index.NoopLogger();
    }
    prepareQuery(query, fields, executeMethod) {
        const stmt = this.client.prepare(query.sql);
        return new PreparedQuery(stmt, query.sql, query.params, this.logger, fields, executeMethod);
    }
    prepareOneTimeQuery(query, fields, executeMethod, customResultMapper) {
        const stmt = this.client.prepare(query.sql);
        return new PreparedQuery(stmt, query.sql, query.params, this.logger, fields, executeMethod, customResultMapper, true);
    }
    transaction(transaction, config = {}) {
        const tx = new SQLJsTransaction('sync', this.dialect, this, this.schema);
        this.run(index.sql.raw(`begin${config.behavior ? ` ${config.behavior}` : ''}`));
        try {
            const result = transaction(tx);
            this.run(index.sql `commit`);
            return result;
        }
        catch (err) {
            this.run(index.sql `rollback`);
            throw err;
        }
    }
}
_a = index.entityKind;
SQLJsSession[_a] = 'SQLJsSession';
class SQLJsTransaction extends session.SQLiteTransaction {
    transaction(transaction) {
        const savepointName = `sp${this.nestedIndex + 1}`;
        const tx = new SQLJsTransaction('sync', this.dialect, this.session, this.schema, this.nestedIndex + 1);
        tx.run(index.sql.raw(`savepoint ${savepointName}`));
        try {
            const result = transaction(tx);
            tx.run(index.sql.raw(`release savepoint ${savepointName}`));
            return result;
        }
        catch (err) {
            tx.run(index.sql.raw(`rollback to savepoint ${savepointName}`));
            throw err;
        }
    }
}
_b = index.entityKind;
SQLJsTransaction[_b] = 'SQLJsTransaction';
class PreparedQuery extends session.PreparedQuery {
    constructor(stmt, queryString, params, logger, fields, executeMethod, customResultMapper, isOneTimeQuery = false) {
        super('sync', executeMethod);
        this.stmt = stmt;
        this.queryString = queryString;
        this.params = params;
        this.logger = logger;
        this.fields = fields;
        this.customResultMapper = customResultMapper;
        this.isOneTimeQuery = isOneTimeQuery;
    }
    run(placeholderValues) {
        const params = index.fillPlaceholders(this.params, placeholderValues ?? {});
        this.logger.logQuery(this.queryString, params);
        const result = this.stmt.run(params);
        if (this.isOneTimeQuery) {
            this.free();
        }
        return result;
    }
    all(placeholderValues) {
        const { fields, joinsNotNullableMap, logger, queryString, stmt, isOneTimeQuery, customResultMapper } = this;
        if (!fields && !customResultMapper) {
            const params = index.fillPlaceholders(this.params, placeholderValues ?? {});
            logger.logQuery(queryString, params);
            stmt.bind(params);
            const rows = [];
            while (stmt.step()) {
                rows.push(stmt.getAsObject());
            }
            if (isOneTimeQuery) {
                this.free();
            }
            return rows;
        }
        const rows = this.values(placeholderValues);
        if (customResultMapper) {
            return customResultMapper(rows, normalizeFieldValue);
        }
        return rows.map((row) => index.mapResultRow(fields, row.map((v) => normalizeFieldValue(v)), joinsNotNullableMap));
    }
    get(placeholderValues) {
        const params = index.fillPlaceholders(this.params, placeholderValues ?? {});
        this.logger.logQuery(this.queryString, params);
        const { fields, stmt, isOneTimeQuery, joinsNotNullableMap, customResultMapper } = this;
        if (!fields && !customResultMapper) {
            const result = stmt.getAsObject(params);
            if (isOneTimeQuery) {
                this.free();
            }
            return result;
        }
        const row = stmt.get(params);
        if (isOneTimeQuery) {
            this.free();
        }
        if (!row || (row.length === 0 && fields.length > 0)) {
            return undefined;
        }
        if (customResultMapper) {
            return customResultMapper([row], normalizeFieldValue);
        }
        return index.mapResultRow(fields, row.map((v) => normalizeFieldValue(v)), joinsNotNullableMap);
    }
    values(placeholderValues) {
        const params = index.fillPlaceholders(this.params, placeholderValues ?? {});
        this.logger.logQuery(this.queryString, params);
        this.stmt.bind(params);
        const rows = [];
        while (this.stmt.step()) {
            rows.push(this.stmt.get());
        }
        if (this.isOneTimeQuery) {
            this.free();
        }
        return rows;
    }
    free() {
        return this.stmt.free();
    }
}
_c = index.entityKind;
PreparedQuery[_c] = 'SQLJsPreparedQuery';
function normalizeFieldValue(value) {
    if (value instanceof Uint8Array) { // eslint-disable-line no-instanceof/no-instanceof
        if (typeof Buffer !== 'undefined') {
            if (!(value instanceof Buffer)) { // eslint-disable-line no-instanceof/no-instanceof
                return Buffer.from(value);
            }
            return value;
        }
        if (typeof TextDecoder !== 'undefined') {
            return new TextDecoder().decode(value);
        }
        throw new Error('TextDecoder is not available. Please provide either Buffer or TextDecoder polyfill.');
    }
    return value;
}

function drizzle(client, config = {}) {
    const dialect = new session.SQLiteSyncDialect();
    let logger;
    if (config.logger === true) {
        logger = new index.DefaultLogger();
    }
    else if (config.logger !== false) {
        logger = config.logger;
    }
    let schema;
    if (config.schema) {
        const tablesConfig = index.extractTablesRelationalConfig(config.schema, index.createTableRelationsHelpers);
        schema = {
            fullSchema: config.schema,
            schema: tablesConfig.tables,
            tableNamesMap: tablesConfig.tableNamesMap,
        };
    }
    const session$1 = new SQLJsSession(client, dialect, schema, { logger });
    return new session.BaseSQLiteDatabase('sync', dialect, session$1, schema);
}

exports.PreparedQuery = PreparedQuery;
exports.SQLJsSession = SQLJsSession;
exports.SQLJsTransaction = SQLJsTransaction;
exports.drizzle = drizzle;
