// drizzle-orm@0.28.6/node-postgres/index.cjs
'use strict';

var pg = require('pg');
var index = require('../index-1899b9ae.cjs');

var _a$1, _b, _c;
const { Pool } = pg;
class NodePgPreparedQuery extends index.PreparedQuery {
    constructor(client, queryString, params, logger, fields, name, customResultMapper) {
        super();
        this.client = client;
        this.params = params;
        this.logger = logger;
        this.fields = fields;
        this.customResultMapper = customResultMapper;
        this.rawQuery = {
            name,
            text: queryString,
        };
        this.query = {
            name,
            text: queryString,
            rowMode: 'array',
        };
    }
    async execute(placeholderValues = {}) {
        return index.tracer.startActiveSpan('drizzle.execute', async () => {
            const params = index.fillPlaceholders(this.params, placeholderValues);
            this.logger.logQuery(this.rawQuery.text, params);
            const { fields, rawQuery, client, query, joinsNotNullableMap, customResultMapper } = this;
            if (!fields && !customResultMapper) {
                return index.tracer.startActiveSpan('drizzle.driver.execute', async (span) => {
                    span?.setAttributes({
                        'drizzle.query.name': rawQuery.name,
                        'drizzle.query.text': rawQuery.text,
                        'drizzle.query.params': JSON.stringify(params),
                    });
                    return client.query(rawQuery, params);
                });
            }
            const result = await index.tracer.startActiveSpan('drizzle.driver.execute', (span) => {
                span?.setAttributes({
                    'drizzle.query.name': query.name,
                    'drizzle.query.text': query.text,
                    'drizzle.query.params': JSON.stringify(params),
                });
                return client.query(query, params);
            });
            return index.tracer.startActiveSpan('drizzle.mapResponse', () => {
                return customResultMapper
                    ? customResultMapper(result.rows)
                    : result.rows.map((row) => index.mapResultRow(fields, row, joinsNotNullableMap));
            });
        });
    }
    all(placeholderValues = {}) {
        return index.tracer.startActiveSpan('drizzle.execute', () => {
            const params = index.fillPlaceholders(this.params, placeholderValues);
            this.logger.logQuery(this.rawQuery.text, params);
            return index.tracer.startActiveSpan('drizzle.driver.execute', (span) => {
                span?.setAttributes({
                    'drizzle.query.name': this.rawQuery.name,
                    'drizzle.query.text': this.rawQuery.text,
                    'drizzle.query.params': JSON.stringify(params),
                });
                return this.client.query(this.rawQuery, params).then((result) => result.rows);
            });
        });
    }
}
_a$1 = index.entityKind;
NodePgPreparedQuery[_a$1] = 'NodePgPreparedQuery';
class NodePgSession extends index.PgSession {
    constructor(client, dialect, schema, options = {}) {
        super(dialect);
        this.client = client;
        this.schema = schema;
        this.options = options;
        this.logger = options.logger ?? new index.NoopLogger();
    }
    prepareQuery(query, fields, name, customResultMapper) {
        return new NodePgPreparedQuery(this.client, query.sql, query.params, this.logger, fields, name, customResultMapper);
    }
    async transaction(transaction, config) {
        const session = this.client instanceof Pool // eslint-disable-line no-instanceof/no-instanceof
            ? new NodePgSession(await this.client.connect(), this.dialect, this.schema, this.options)
            : this;
        const tx = new NodePgTransaction(this.dialect, session, this.schema);
        await tx.execute(index.sql `begin${config ? index.sql ` ${tx.getTransactionConfigSQL(config)}` : undefined}`);
        try {
            const result = await transaction(tx);
            await tx.execute(index.sql `commit`);
            return result;
        }
        catch (error) {
            await tx.execute(index.sql `rollback`);
            throw error;
        }
        finally {
            if (this.client instanceof Pool) { // eslint-disable-line no-instanceof/no-instanceof
                session.client.release();
            }
        }
    }
}
_b = index.entityKind;
NodePgSession[_b] = 'NodePgSession';
class NodePgTransaction extends index.PgTransaction {
    async transaction(transaction) {
        const savepointName = `sp${this.nestedIndex + 1}`;
        const tx = new NodePgTransaction(this.dialect, this.session, this.schema, this.nestedIndex + 1);
        await tx.execute(index.sql.raw(`savepoint ${savepointName}`));
        try {
            const result = await transaction(tx);
            await tx.execute(index.sql.raw(`release savepoint ${savepointName}`));
            return result;
        }
        catch (err) {
            await tx.execute(index.sql.raw(`rollback to savepoint ${savepointName}`));
            throw err;
        }
    }
}
_c = index.entityKind;
NodePgTransaction[_c] = 'NodePgTransaction';

var _a;
const { types } = pg;
class NodePgDriver {
    constructor(client, dialect, options = {}) {
        this.client = client;
        this.dialect = dialect;
        this.options = options;
        this.initMappers();
    }
    createSession(schema) {
        return new NodePgSession(this.client, this.dialect, schema, { logger: this.options.logger });
    }
    initMappers() {
        types.setTypeParser(types.builtins.TIMESTAMPTZ, (val) => val);
        types.setTypeParser(types.builtins.TIMESTAMP, (val) => val);
        types.setTypeParser(types.builtins.DATE, (val) => val);
        types.setTypeParser(types.builtins.INTERVAL, (val) => val);
    }
}
_a = index.entityKind;
NodePgDriver[_a] = 'NodePgDriver';
function drizzle(client, config = {}) {
    const dialect = new index.PgDialect();
    let logger;
    if (config.logger === true) {
        logger = new index.DefaultLogger();
    }
    else if (config.logger !== false) {
        logger = config.logger;
    }
    let schema;
    if (config.schema) {
        const tablesConfig = index.extractTablesRelationalConfig(config.schema, index.createTableRelationsHelpers);
        schema = {
            fullSchema: config.schema,
            schema: tablesConfig.tables,
            tableNamesMap: tablesConfig.tableNamesMap,
        };
    }
    const driver = new NodePgDriver(client, dialect, { logger });
    const session = driver.createSession(schema);
    return new index.PgDatabase(dialect, session, schema);
}

exports.NodePgDriver = NodePgDriver;
exports.NodePgPreparedQuery = NodePgPreparedQuery;
exports.NodePgSession = NodePgSession;
exports.NodePgTransaction = NodePgTransaction;
exports.drizzle = drizzle;
