// drizzle-orm@0.28.6/better-sqlite3/index.cjs
'use strict';

var index = require('../index-1899b9ae.cjs');
var session = require('../session-90f70450.cjs');

var _a, _b, _c;
class BetterSQLiteSession extends session.SQLiteSession {
    constructor(client, dialect, schema, options = {}) {
        super(dialect);
        this.client = client;
        this.schema = schema;
        this.logger = options.logger ?? new index.NoopLogger();
    }
    prepareQuery(query, fields, executeMethod, customResultMapper) {
        const stmt = this.client.prepare(query.sql);
        return new PreparedQuery(stmt, query.sql, query.params, this.logger, fields, executeMethod, customResultMapper);
    }
    transaction(transaction, config = {}) {
        const tx = new BetterSQLiteTransaction('sync', this.dialect, this, this.schema);
        const nativeTx = this.client.transaction(transaction);
        return nativeTx[config.behavior ?? 'deferred'](tx);
    }
}
_a = index.entityKind;
BetterSQLiteSession[_a] = 'BetterSQLiteSession';
class BetterSQLiteTransaction extends session.SQLiteTransaction {
    transaction(transaction) {
        const savepointName = `sp${this.nestedIndex}`;
        const tx = new BetterSQLiteTransaction('sync', this.dialect, this.session, this.schema, this.nestedIndex + 1);
        this.session.run(index.sql.raw(`savepoint ${savepointName}`));
        try {
            const result = transaction(tx);
            this.session.run(index.sql.raw(`release savepoint ${savepointName}`));
            return result;
        }
        catch (err) {
            this.session.run(index.sql.raw(`rollback to savepoint ${savepointName}`));
            throw err;
        }
    }
}
_b = index.entityKind;
BetterSQLiteTransaction[_b] = 'BetterSQLiteTransaction';
class PreparedQuery extends session.PreparedQuery {
    constructor(stmt, queryString, params, logger, fields, executeMethod, customResultMapper) {
        super('sync', executeMethod);
        this.stmt = stmt;
        this.queryString = queryString;
        this.params = params;
        this.logger = logger;
        this.fields = fields;
        this.customResultMapper = customResultMapper;
    }
    run(placeholderValues) {
        const params = index.fillPlaceholders(this.params, placeholderValues ?? {});
        this.logger.logQuery(this.queryString, params);
        return this.stmt.run(...params);
    }
    all(placeholderValues) {
        const { fields, joinsNotNullableMap, queryString, logger, stmt, customResultMapper } = this;
        if (!fields && !customResultMapper) {
            const params = index.fillPlaceholders(this.params, placeholderValues ?? {});
            logger.logQuery(queryString, params);
            return stmt.all(...params);
        }
        const rows = this.values(placeholderValues);
        if (customResultMapper) {
            return customResultMapper(rows);
        }
        return rows.map((row) => index.mapResultRow(fields, row, joinsNotNullableMap));
    }
    get(placeholderValues) {
        const params = index.fillPlaceholders(this.params, placeholderValues ?? {});
        this.logger.logQuery(this.queryString, params);
        const { fields, stmt, joinsNotNullableMap, customResultMapper } = this;
        if (!fields && !customResultMapper) {
            return stmt.get(...params);
        }
        const row = stmt.raw().get(...params);
        if (!row) {
            return undefined;
        }
        if (customResultMapper) {
            return customResultMapper([row]);
        }
        return index.mapResultRow(fields, row, joinsNotNullableMap);
    }
    values(placeholderValues) {
        const params = index.fillPlaceholders(this.params, placeholderValues ?? {});
        this.logger.logQuery(this.queryString, params);
        return this.stmt.raw().all(...params);
    }
}
_c = index.entityKind;
PreparedQuery[_c] = 'BetterSQLitePreparedQuery';

function drizzle(client, config = {}) {
    const dialect = new session.SQLiteSyncDialect();
    let logger;
    if (config.logger === true) {
        logger = new index.DefaultLogger();
    }
    else if (config.logger !== false) {
        logger = config.logger;
    }
    let schema;
    if (config.schema) {
        const tablesConfig = index.extractTablesRelationalConfig(config.schema, index.createTableRelationsHelpers);
        schema = {
            fullSchema: config.schema,
            schema: tablesConfig.tables,
            tableNamesMap: tablesConfig.tableNamesMap,
        };
    }
    const session$1 = new BetterSQLiteSession(client, dialect, schema, { logger });
    return new session.BaseSQLiteDatabase('sync', dialect, session$1, schema);
}

exports.BetterSQLiteSession = BetterSQLiteSession;
exports.BetterSQLiteTransaction = BetterSQLiteTransaction;
exports.PreparedQuery = PreparedQuery;
exports.drizzle = drizzle;
